client/src/main/java/org/eclipse/hono/client/impl/HonoConnectionImpl.java

```diff
@@ -708,11 +709,16 @@ public void closeAndFree(
                 receiver.setQoS(qos);
                 receiver.setPrefetch(preFetchSize);
                 receiver.handler((delivery, message) -> {
-                    messageHandler.handle(delivery, message);
-                    if (log.isTraceEnabled()) {
-                        final int remainingCredits = receiver.getCredit() - receiver.getQueued();
-                        log.trace("handling message [remotely settled: {}, queued messages: {}, remaining credit: {}]",
-                                delivery.remotelySettled(), receiver.getQueued(), remainingCredits);
+                    try {
+                        messageHandler.handle(delivery, message);
+                        if (log.isTraceEnabled()) {
+                            final int remainingCredits = receiver.getCredit() - receiver.getQueued();
+                            log.trace("handling message [remotely settled: {}, queued messages: {}, remaining credit: {}]",
+                                    delivery.remotelySettled(), receiver.getQueued(), remainingCredits);
+                        }
+                    } catch (final Exception ex) {
+                        log.warn("error handling message", ex);
+                        ProtonHelper.modified(delivery, true, true, false);
```

Comment:
I'd rather release the message here since we have no clue why the processing failed and we cannot indicate to the sender, what needs to be changed in order for the next attempt to succeed.

Reply:
Right, I've changed that accordingly.